### PR TITLE
JAMES-3715 Several Netty 4 fixes and improvements to make starting and managing netty servers easier

### PR DESCRIPTION
**Pull request by glennosss**
Several fixes and improvements to make starting and managing netty servers easier - specifically:
1. Created 'EventLoopGroupManager' interface - see interface javadoc for more information - allows much better control over starting/sharing EventLoopGroups - https://github.com/glennosss/james-project/blob/netty4-combo/protocols/netty/src/main/java/org/apache/james/protocols/netty/EventLoopGroupManager.java
2. Updated 'Encryption' to support old static methods, but added additional static methods which add Netty SslContext support - see 'Encryption' javadoc for more information - https://github.com/glennosss/james-project/blob/netty4-combo/protocols/api/src/main/java/org/apache/james/protocols/api/Encryption.java
3. See 'ProtocolServer' javadoc - now supports async bind/unbind, control over how long unbind takes, and added unbindNow() to help with instantly stopping server for unit tests - https://github.com/glennosss/james-project/blob/netty4-combo/protocols/api/src/main/java/org/apache/james/protocols/api/ProtocolServer.java
4. Updated 'NettyServer', 'AbstractAsyncServer' to support proper construction using Factory/Builder approach - making it much easier to method chain and build a server.
5. Updated 'AbstractConfigurableAsyncServer' to support Factory/Builder approach - also 'bossThreads' and 'keepalive' parameters are now configurable.

FYI Also includes a fix for TCP_NODELAY still wasn't being set properly (`configureBootstrap()` was being run after socket was bound).

### CODE REVIEW

**Comment by glennosss** [comment]
This pull request replaces https://github.com/chibenwa/james-project/pull/31

File: protocols/netty/src/main/java/org/apache/james/protocols/netty/AbstractAsyncServer.java [hunk 1]
```diff
@@ -93,7 +93,7 @@ public synchronized void bind() throws Exception {
         ServerBootstrap bootstrap = new ServerBootstrap();
         bootstrap.channel(NioServerSocketChannel.class);
 
-        bossGroup = new NioEventLoopGroup(NamedThreadFactory.withName(jmxName + "-boss"));
+        bossGroup = new NioEventLoopGroup(2, NamedThreadFactory.withName(jmxName + "-boss"));
```
Comment (glennosss) [inline protocols/netty/src/main/java/org/apache/james/protocols/netty/AbstractAsyncServer.java]: FYI These types of hard coded thread pool sizes are a big part of what this pull request is trying to fix.

**Review by glennosss — COMMENTED** [review]
Also created 'EventLoopGroupManager' interface to fix issues like this one:

**Comment by glennosss** [comment]
@chibenwa What's involved in getting this pull request approved and merged?

**Comment by glennosss** [comment]
A shared 'EventLoopGroupManager' is also an important part of one doing being able to support a NIO client: https://issues.apache.org/jira/browse/JAMES-964

**Comment by chibenwa** [comment]
This work is not up to date with master. Could you rebase?

File: protocols/api/pom.xml [hunk 2]
```diff
@@ -46,6 +46,10 @@
             <groupId>com.google.guava</groupId>
             <artifactId>guava</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-handler</artifactId>
+        </dependency>
```
Comment (chibenwa) [inline protocols/api/pom.xml:52]: I think this dependency was there on purpose, not to hard code the use of netty in the middle of protocols-api

File: protocols/api/src/main/java/org/apache/james/protocols/api/Encryption.java [hunk 3]
```diff
@@ -19,26 +19,41 @@
...
+ * 
+ * Netty {@link SslContext} can be configured to use Google's BoringSSL.
+ * BoringSSL is a clone of OpenSSL with all redundant/old ciphers removed and is compiled into to all the Chrome/Chromium browsers.
+ * Simply include the 'io.netty:netty-tcnative-boringssl-static' dependency as discussed here: https://netty.io/wiki/forked-tomcat-native.html
+ * You should regularly update the version of this library to keep in-sync with the latest version used by Chrome/Chromium browsers.
+ * 
+ * Use {@link SslContextBuilder} to create an instance of {@link SslContext}.
  */
-public final class Encryption {
+public abstract class Encryption {
```
Comment (chibenwa) [inline protocols/api/src/main/java/org/apache/james/protocols/api/Encryption.java:47]: I worked on a clean abstraction, turning Encryption into an interface, moving it to protocols-netty, and providing Netty implementation of it.

Though I failed at using Netty native bindings... netty-ctnative was built against openssl-1.0 while I have openssl-1.1... Feeling sad here!

I'll PR this generalisation of OpenSSL separately...

File: protocols/netty/src/main/java/org/apache/james/protocols/netty/LineDelimiterBasedChannelHandlerFactory.java [hunk 4]
```diff
@@ -32,7 +32,7 @@ public LineDelimiterBasedChannelHandlerFactory(int maxLineLength) {
 
     @Override
     public ChannelHandler create(ChannelPipeline pipeline) {
-        return new LineBasedFrameDecoder(maxLineLength, false, !FAIL_FAST);
+        return new LineBasedFrameDecoder(maxLineLength, false, FAIL_FAST);
```
Comment (chibenwa) [inline protocols/netty/src/main/java/org/apache/james/protocols/netty/LineDelimiterBasedChannelHandlerFactory.java:35]: Unrelated change, and the constant name is no longer appropriate.

File: protocols/netty/src/main/java/org/apache/james/protocols/netty/EventLoopGroupManagerDefault.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+import io.netty.util.concurrent.EventExecutorGroup;
+import io.netty.util.concurrent.PromiseCombiner;
+
+public final class EventLoopGroupManagerDefault implements EventLoopGroupManager {
+
+    private static final int PROCESSORS = Runtime.getRuntime().availableProcessors();
+
+    public static final int DEFAULT_BOSS_THREADS = Math.min(2, PROCESSORS);
+    public static final int DEFAULT_WORKER_THREADS = PROCESSORS * 2;
+    public static final int DEFAULT_EXECUTOR_GROUP_THREADS = PROCESSORS * 2; // TODO potentially recommend setting this to 0 so reuses worker thread pool.
```
Comment (chibenwa) [inline protocols/netty/src/main/java/org/apache/james/protocols/netty/EventLoopGroupManagerDefault.java:39]: We can't as we have blocking logic that should not be executed on the IO eventloop...

File: protocols/api/src/main/java/org/apache/james/protocols/api/ProtocolServer.java [hunk 6]
```diff
@@ -46,26 +46,26 @@
      * Start the server - blocks until server has started.
      */
     default void bind() {
-        bindAsync().syncUninterruptibly();
+        sync_bind(bindAsync());
```
Comment (chibenwa) [inline protocols/api/src/main/java/org/apache/james/protocols/api/ProtocolServer.java:49]: syncBind

**Review by chibenwa — COMMENTED** [review]
To be fairly honnest this PR adds quite a bunch of new abstractions, I fail at seeing the point of this EventLoopGroupManager that in my view overlaps with the reponsibility of AbstractAsyncServer.

**Comment by chibenwa** [comment]
Here is what I came up with regarding encryption: https://github.com/apache/james-project/pull/930

**Comment by glennosss** [comment]
I'm having to move onto other projects. So good luck with the project guys. I'll close all my pull requests. But feel free to take whatever parts you did like if you want and do whatever you want with them. Later all!
